basic_features/basic_mod_data_checker.py

```diff
@@ -0,0 +1,258 @@
+        )
+
+
+class GlobPatternDict(TypedDict, total=False):
+    """See: `BasicModDataChecker`"""
+
+    unfold: Iterable[str] | None
+    valid: Iterable[str] | None
+    delete: Iterable[str] | None
+    move: MutableMapping[str, str] | None
+
+
+class BasicModDataChecker(mobase.ModDataChecker):
+    """Game feature that is used to check and fix the content of a data tree
+    via simple file definitions.
+
+    The file definitions support glob pattern (without subfolders) and are checked in
+    definition order (passed either as dict or kwargs).
+
+    Args:
+        file_patterns (optional): An dict with the keys below or as kwargs:
+
+        unfold (optional): Folders to unfold (remove and move contents to parent),
+            after being checked and fixed recursively.
+
+        valid (optional): Files and folders in the right path.
+            Check result: `mobase.ModDataChecker.VALID`.
+
+        delete (optional): Files/folders to delete.
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        move (optional): Files/folders to move and their target.
+            If the path ends with `/` or `\\`, the entry will be inserted
+            in the corresponding directory instead of replacing it.
+
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        Example::
+
+            BasicModDataChecker(
+                valid=["valid_folder", "*.ext1"]
+                move={"*.ext2": "path/to/target_folder/"}
+            )
+
+            BasicModDataChecker(
+                {
+                    "valid": ["valid_folder", "*.ext1"],
+                    "move": {"*.ext2": "path/to/target_folder/"}
+                }
+            )
+
+    See Also:
+        `mobase.IFileTree.move`
+    """
+
+    file_patterns = None  # type: GlobPatternDict
+    """Use `update_patterns` for modifications."""
+
+    _regex: RegexPatternDict
+    """The regex patterns derived from the file (glob) patterns."""
+
+    _move_targets: Sequence[str]
+
+    # Overloads as workaround for **kwargs: Unpack[GlobPatternDict]
+    @overload
+    def __init__(
+        self,
+        *,
+        unfold: Iterable[str] | None = None,
+        valid: Iterable[str] | None = None,
+        delete: Iterable[str] | None = None,
+        move: dict[str, str] | None = None,
+    ):
+        ...
+
+    @overload
+    def __init__(
+        self,
+        file_patterns: Optional[GlobPatternDict] = None,
+        *,
+        unfold: Iterable[str] | None = None,
+        valid: Iterable[str] | None = None,
+        delete: Iterable[str] | None = None,
+        move: dict[str, str] | None = None,
+    ):
+        ...
+
+    def __init__(
+        self, file_patterns: Optional[GlobPatternDict] = None, **kwargs
+    ):  # Unpack[GlobPatternDict]
+        super().__init__()
+        # Init with copy from class var
+        self.file_patterns = fp.copy() if (fp := self.file_patterns) else {}
```

Comment:
I don't understand what you're trying to do here... Either this is fishy or you need to add a more concrete explanation as why you would want to copy the class var here.

Reply:
If the class var is defined as a mutable field, it would be shared between all instances, which could lead to unexpected results (see comment above). That is why I am making sure, that the dict is new for each instance.

Reply:
I don't understand why you need to copy `self.file_patterns`... Basically you would like to allow child classes to define a class-wide `file_patterns` than is then copied to all instances? If yes, then use two different names: one for the class instance, e.g., `default_file_patterns`, that you initialize to `{}` instead of `None` so that you can do proper typing and then `file_patterns` (or `_file_patterns` the instance one). If it's not what you want to do, I still don't understand.

Reply:
Ok I can think of 3 scenarios, which I like to cover:
1.  `BasicModDataChecker(file_patterns=...)` gets called in `BasicGame.init()` with the file patterns directly defined.
(like in [Valheim @ line 329-327](90/files#diff-148017f5fb4a148be37130e90abedfabb2cf66ccc24d187c035689275ad2ad83R329-R374))
  **Pro:** no subclass
  **Con:** cannot be imported to other games.
2. Defining a subclass and use it for multiple games for e.g. in the updates for [Subnautica ](/ZashIn/modorganizer-basic_games/blob/1cea78a8ee52480c7230af6af6201b0dd181c793/games/game_subnautica.py#L20-L32) and [Subnautica Below Zero](/ZashIn/modorganizer-basic_games/blob/1cea78a8ee52480c7230af6af6201b0dd181c793/games/game_subnautica-below-zero.py#L32):
    ```py
    SubnauticaModDataChecker(BasicModDataChecker):
        file_pattern = {...}
    ````
    **Pro:** can be used my multiple games
    **Con:** subclass
    I would avoid using the same instance in multiple games...
3. Add custom checks / fixes in the subclass, by overwriting `.dataLooksValid()` and `.fix()`.
    Defining both file patterns and the custom behavior together in the class, makes it a lot easier to understand and read,
    in contrast to defining the custom checks in the class body and the `file_pattern` in the `BasicGame.init()`

Reply:
2) can be done by doing

```python
class SubnauticaModDataChecker(BasicModDataChecker):
    
    def __init__(self, file_patterns):
        super().__init__({
        # some default stuff for this class
        } | file_patterns)
```

avoiding the need for a class attribute , if you really want a class attribute, at least use a different name as I commented.

Reply:
Right that works, too, Although `|` does not work with a dict. And one has to be careful with the shared objects, since `file_patterns: ... = {}` in the `__init__` would use the same dict object for all instances as default.

Same goes for `BasicModDataChecker(SubnauticaModDataChecker.file_pattern)).update_filepatterns(...)`:
Without making a copy in init, this would update the class field `file pattern` (or e.g. `default_file_pattern`).

The question is when to overwrite, and when to modify the existing / default file patterns.
The clearest distinction would probably be to overwrite in `__init__` and via setter, but to update via `update_patterns`.

Something like this (without the kwargs):
```py
class BasicModDataChecker:
    default_file_patterns: ClassVar[Optional[GlobPatternDict]] = None
    _file_patterns: GlobPatternDict
    
    def __init__(self, file_patterns: Optional[GlobPatternDict] = None):
        super().__init__()
        # Init with copy from class var by default (prevents modification of class var)
        self.file_patterns = file_patterns or (
            dfp.copy() if (dfp := self.default_file_patterns) else {}
        )

    @property
    def file_patterns(self):
        return self._file_patterns

    @file_patterns.setter
    def file_patterns(self, file_patterns: GlobPatternDict):
        self._file_patterns = file_patterns
        self.update_patterns()

    def update_patterns(self, file_patterns: Optional[GlobPatternDict]):
        if file_patterns:
            self._file_patterns.update(file_patterns)
        # Update other variables
        ...

class SubnauticaModDataChecker(BasicModDataChecker):
    default_file_patterns = ...
```

Or without the class var:
```py
class BasicModDataChecker:
    _file_patterns: GlobPatternDict
    
    def __init__(self, file_patterns: GlobPatternDict):
        super().__init__()
        self.file_patterns = file_patterns  
    ...

class SubnauticaModDataChecker(BasicModDataChecker):
    def __init__(self, file_patterns: Optional[GlobPatternDict] = None):
        super().__init__(
            file_patterns
            or {
                # default patterns
            }
        )
```

And using the subclass for another game (with both versions above):
```py
BelowZeroModDataChecker = SubnauticaModDataChecker()
BelowZeroModDataChecker.update_file_patterns({...}) #  Preserves existing order of definitions / actions
```

Reply:
The first solution looks good to me, although 

1) is there any needs for defaulting `default_file_patterns` to `None` instead of `{}` since you copy anyway?

2) I would not expose `file_patterns`, simply keep it has `_file_patterns` and do the `update()` in `__init__`, no needs for property.

Reply:
> 2\. I would not expose `file_patterns`, simply keep it has `_file_patterns` and do the `update()` in `__init__`, no needs for property.
Without a setter / `.overwrite_file_pattern()`, that would mean you cannot change the order of the actions after initializing the object though (since you cannot copy the interal `_file_patterns` either).

So: no property and an extra method `overwrite_file_pattern()`?

Reply:
Added `set_patterns`.
> 1. is there any needs for defaulting `default_file_patterns` to `None` instead of `{}` since you copy anyway?
Basically yes, done it now in f3977a9. Only remark would be if someone adds to the default patterns, instead of overwriting them, but since it is a `ClassVar` now, that should be clear to everybody.